### PR TITLE
Implement some of TerrainLogic

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/terrainroads.h [hunk 1]
```diff
@@ -62,6 +62,7 @@ class TerrainRoadType : public MemoryPoolObject
     Utf8String Get_Bridge_Model_Name_Really_Damaged() { return m_bridgeModelNameReallyDamaged; }
     Utf8String Get_Texture_Broken() { return m_textureBroken; }
     Utf8String Get_Bridge_Model_Name_Broken() { return m_bridgeModelNameBroken; }
+    Utf8String Get_Tower_Object_Name(BridgeTowerType type) { return m_towerObjectName[type]; }
```
Comment (xezon) [inline src/game/client/terrainroads.h:65]: Can be const

File: src/game/common/system/geometry.h [hunk 2]
```diff
@@ -65,6 +65,7 @@ class GeometryInfo : public SnapShot
     float Get_Bounding_Circle_Radius() const { return m_boundingCircleRadius; }
     float Get_Bounding_Sphere_Radius() const { return m_boundingSphereRadius; }
     GeometryType Get_Type() const { return m_type; }
+    bool Is_Small() const { return m_isSmall; }
```
Comment (xezon) [inline src/game/common/system/geometry.h:68]: Can be const

File: src/game/logic/map/terrainlogic.h [hunk 3]
```diff
@@ -89,6 +89,9 @@ class Bridge : public MemoryPoolObject
     Bridge *Get_Next() { return m_next; }
     void Set_Layer(PathfindLayerEnum layer) { m_layer = layer; }
     PathfindLayerEnum Get_Layer() { return m_layer; }
+    Region2D Get_Bounds() { return m_bounds; }
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.h:92]: Can be const 2x

File: src/game/logic/map/terrainlogic.h [hunk 4]
```diff
@@ -221,10 +242,16 @@ class TerrainLogic : public SnapShot, public SubsystemInterface
     PathfindLayerEnum Get_Layer_For_Destination(const Coord3D *pos);
     PathfindLayerEnum Get_Highest_Layer_For_Destination(const Coord3D *pos, bool b);
 
+    bool Have_Bridge_Damage_States_Changed() { return m_bridgeDamageStatesChanged; }
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.h:245]: Can be const

File: src/game/logic/map/terrainlogic.cpp [hunk 5]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+        bridge->Set_Position(&pos);
+        m_bridgeInfo.bridge_object_id = bridge->Get_ID();
+        bridge->Update_Obj_Values_From_Map_Properties(props);
+        Coord2D c;
+        c.x = m_bridgeInfo.to_left.x - m_bridgeInfo.from_left.x;
+        c.y = m_bridgeInfo.to_left.y - m_bridgeInfo.from_left.y;
+        bridge->Set_Orientation(c.To_Angle());
+
+        if (g_theTerrainRoads->Find_Bridge(bridge_template_name)) {
+            m_next = nullptr;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:225]: Not `m_next = bridge`? Looks like `m_next` is never assigned anything other than nullptr in the current implementation. This looks odd.

File: src/game/logic/map/terrainlogic.cpp [hunk 6]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    p1.y = from_right.y;
+
+    p2.x = to_right.x;
+    p2.y = to_right.y;
+
+    if (Line_In_Region(&p1, &p2, cell)) {
+        return true;
+    }
+
+    from_left.x = from_left.x - c.x;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: Or +=, -=, 8x

File: src/game/logic/map/terrainlogic.cpp [hunk 7]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    c.x = c.x * 10.0f;
+    c.y = c.y * 10.0f;
+
+    Coord3D c2 = m_bridgeInfo.to - m_bridgeInfo.from;
+    c2.Normalize();
+    c2.x = c2.x * 5.0f;
+    c2.y = c2.y * 5.0f;
+
+    Coord3D from_left = m_bridgeInfo.from_left;
+    from_left.x = from_left.x - c2.x;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: Or +=, -=, 16x

File: src/game/logic/map/terrainlogic.cpp [hunk 8]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    }
+
+    return false;
+}
+
+bool Bridge::Is_Cell_Entry_Point(const Region2D *cell)
+{
+    Coord3D c = m_bridgeInfo.from_right - m_bridgeInfo.from_left;
+    c.Normalize();
+    c.x = c.x * 10.0f;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: Or *= 4x

File: src/game/logic/map/terrainlogic.cpp [hunk 9]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+}
+
+bool TerrainLogic::Is_Cliff_Cell(float x, float y) const
+{
+    return false;
+}
+
+bool TerrainLogic::Parse_Waypoint_Data_Chunk(DataChunkInput &file, DataChunkInfo *info, void *user_data)
+{
+    return (static_cast<TerrainLogic *>(user_data))->Parse_Waypoint_Data(file, info, user_data);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: Can remove the () around static_cast

File: src/game/logic/map/terrainlogic.cpp [hunk 10]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+                    distance = dist;
+                }
+            }
+        }
+
+        return waypoint;
+    }
+}
+
+bool TerrainLogic::Is_Purpose_Of_Path(Waypoint *way, Utf8String label)
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1023]: waypoint

File: src/game/logic/map/terrainlogic.cpp [hunk 11]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+        }
+
+        return false;
+    } else {
+        captainslog_debug("***Warning - asking for empth path label.");
+        return false;
+    }
+}
+
+Bridge *TerrainLogic::Find_Bridge_At(const Coord3D *loc) const
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1045]: const Coord3D *pos

File: src/game/logic/map/terrainlogic.cpp [hunk 12]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    for (Bridge *bridge = Get_First_Bridge(); bridge != nullptr; bridge = bridge->Get_Next()) {
+        if (bridge->Is_Point_On_Bridge(loc)) {
+            return bridge;
+        }
+    }
+
+    return nullptr;
+}
+
+Bridge *TerrainLogic::Find_Bridge_Layer_At(const Coord3D *loc, PathfindLayerEnum layer, bool b) const
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1056]: const Coord3D *pos
bool require_point_on_bridge

if the `if (bridge->Get_Layer() == layer && (!b || bridge->Is_Point_On_Bridge(loc)))` condition is correct.

File: src/game/logic/map/terrainlogic.cpp [hunk 13]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    for (Bridge *bridge = Get_First_Bridge(); bridge != nullptr; bridge = bridge->Get_Next()) {
+        if (bridge->Get_Layer() == layer && (!b || bridge->Is_Point_On_Bridge(loc))) {
+            return bridge;
+        }
+    }
+
+    return nullptr;
+}
+
 PathfindLayerEnum TerrainLogic::Get_Highest_Layer_For_Destination(const Coord3D *pos, bool b)
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1071]: bool require_intact_bridge

if the `if ((!b || bridge->Peek_Bridge_Info()->cur_damage_state != BODY_RUBBLE) && bridge->Is_Point_On_Bridge(pos))` condition is correct.

File: src/game/logic/map/terrainlogic.cpp [hunk 14]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
-#else
-    return LAYER_INVALID;
-#endif
+    PathfindLayerEnum layer = LAYER_GROUND;
+
+    float height = pos->z - Get_Ground_Height(pos->x, pos->y, nullptr);
+
+    if (g_theAI->Get_Pathfinder()->Get_Wall_Height() / 2.0f < height) {
+        if (g_theAI->Get_Pathfinder()->Is_Point_On_Wall(pos)) {
+            float wallheight = pos->z - g_theAI->Get_Pathfinder()->Get_Wall_Height();
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: height_above_wall

File: src/game/logic/map/terrainlogic.cpp [hunk 15]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+                    layer = LAYER_WALLS;
+                    height = wallheight;
+                }
+            }
+        }
+    }
+
+    for (Bridge *bridge = Get_First_Bridge(); bridge != nullptr; bridge = bridge->Get_Next()) {
+        if ((!b || bridge->Peek_Bridge_Info()->cur_damage_state != BODY_RUBBLE) && bridge->Is_Point_On_Bridge(pos)) {
+            float bridgeheight = pos->z - bridge->Get_Bridge_Height(pos, nullptr);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: height_above_bridge

File: src/game/logic/map/terrainlogic.cpp [hunk 16]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+                    height = bridgeheight;
+                }
+            }
+        }
+    }
+
+    return layer;
+}
+
+bool TerrainLogic::Object_Interacts_With_Bridge_Layer(Object *obj, int layer, bool b) const
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1106]: PathfindLayerEnum layer?

bool require_intact_bridge

if the `return GameMath::Fabs(obj->Get_Position()->z - height) <= 10.0f
                || (!b || bridge->Peek_Bridge_Info()->cur_damage_state != BODY_RUBBLE)` condition is correct.

File: src/game/logic/map/terrainlogic.cpp [hunk 17]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+            float height = bridge->Get_Bridge_Height(obj->Get_Position(), nullptr);
+            return GameMath::Fabs(obj->Get_Position()->z - height) <= 10.0f
+                || (!b || bridge->Peek_Bridge_Info()->cur_damage_state != BODY_RUBBLE);
+        } else {
+            return false;
+        }
+    }
+}
+
+bool TerrainLogic::Object_Interacts_With_Bridge_End(Object *obj, int layer) const
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1160]: PathfindLayerEnum layer?

File: src/game/logic/map/terrainlogic.cpp [hunk 18]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+    }
+
+    bool found = false;
+
+    float radius = obj->Get_Geometry_Info().Get_Minor_Radius() + 5.0f;
+    Region2D cell;
+    cell.lo.x = obj->Get_Position()->x;
+    cell.lo.y = obj->Get_Position()->y;
+    cell.hi = cell.lo;
+    cell.lo.x = cell.lo.x - radius;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1184]: Or -=, +=, 4x

File: src/game/logic/map/terrainlogic.cpp [hunk 19]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
+{
+    if (bridge == nullptr) {
+        return false;
+    }
+
+    ObjectID id = bridge->Get_ID();
+
+    for (Bridge *bridge = Get_First_Bridge(); bridge != nullptr; bridge = bridge->Get_Next()) {
+        if (bridge == nullptr) {
+            return 0;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: false

File: src/game/logic/map/terrainlogic.cpp [hunk 20]
```diff
@@ -34,14 +52,1205 @@ BridgeInfo::BridgeInfo() :
...
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: false

File: src/game/logic/map/terrainlogic.cpp [hunk 21]
```diff
@@ -64,3 +1273,147 @@ void TerrainLogic::Get_Bridge_Attack_Points(const Object *bridge, TBridgeAttackI
...
+    }
+}
+
+void TerrainLogic::CRC_Snapshot(Xfer *xfer) {}
+
+void TerrainLogic::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 2;
+    xfer->xferVersion(&version, 2);
+    int boundary = m_activeBoundary;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1372]: `int boundary` is not used.

File: src/game/logic/map/terrainlogic.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp:225]: Its set by calls to Bridge::Set_Next
And this code is correct.

File: src/game/logic/map/terrainlogic.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp:1160]: int is correct here

File: src/game/logic/map/terrainlogic.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp:1106]: int is correct here

File: src/game/logic/map/terrainlogic.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp:1184]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp:1372]: fixed.

**Comment by xezon** [comment]
Driving on bridge does not work correctly. Unit is able to pass, but will always cross it in its entirety. It is not possible to send unit half way over bridge. The mouse cursor also incorrectly indicates that the bridge is not passable.

Tested on Sand Serpent (2)

shot_20230417_230543_1

**Comment by xezon** [comment]
AI troops no longer capture civilian structures. Tested on Heartland Shield (2)

shot_20230417_232320_2

**Comment by jonwil** [comment]
Bridge is fixed.
Tested with previous revision, AI doesn't capture oil derricks. Possibly related to #825 or #882

File: src/game/logic/map/terrainlogic.cpp [hunk 22]
```diff
@@ -64,3 +1271,147 @@ void TerrainLogic::Get_Bridge_Attack_Points(const Object *bridge, TBridgeAttackI
...
+            g_theGameLogic->Destroy_Object(obj);
+        }
+
+        bridge->Delete_Instance();
+    }
+}
+
+void TerrainLogic::Set_Active_Boundary(int new_active_boundary)
+{
+    if (new_active_boundary >= 0 && new_active_boundary < m_boundaries.size() && new_active_boundary != m_activeBoundary
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: `src\game\logic\map\terrainlogic.cpp(1335,57): warning C4018: '<': signed/unsigned mismatch`

Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Shell map looks as usual.
Bridges work as original.
Terrain looks and acts normal.

**Pull request merged**